File: src/lib/constants/status.ts
import type { VisitStatus } from "@/types";

export const SCHEDULE_STATUSES: VisitStatus[] = [
  "pending",
  "in_progress",
  "gagal_partial",
  "completed",
  "gagal_total",
];

export const STATUS_TRANSITIONS: Record<VisitStatus, VisitStatus[]> = {
  pending: ["in_progress", "gagal_partial", "completed", "gagal_total"],
  in_progress: ["gagal_partial", "completed", "gagal_total"],
  gagal_partial: ["completed", "gagal_total"],
  completed: ["in_progress"],
  gagal_total: [],
};

export const STATUS_LABELS: Record<VisitStatus, string> = {
  pending: "Pending",
  in_progress: "In Progress",
  gagal_partial: "Gagal Partial",
  completed: "Completed",
  gagal_total: "Gagal Total",
};

export const STATUS_COLORS: Record<VisitStatus, string> = {
  pending: "text-amber-500 bg-amber-50 border-amber-200",
  in_progress: "text-yellow-600 bg-yellow-50 border-yellow-200",
  gagal_partial: "text-orange-500 bg-orange-50 border-orange-200",
  completed: "text-green-500 bg-green-50 border-green-200",
  gagal_total: "text-red-500 bg-red-50 border-red-200",
};

export const STATUS_DOT_COLORS: Record<VisitStatus, string> = {
  pending: "bg-amber-500",
  in_progress: "bg-yellow-500",
  gagal_partial: "bg-orange-500",
  completed: "bg-green-500",
  gagal_total: "bg-red-500",
};

File: fvms_flutter/lib/core/constants/status.dart
/// Mirror src/lib/constants/status.ts
enum VisitStatus {
  pending,
  inProgress,
  gagalPartial,
  completed,
  gagalTotal,
}

extension VisitStatusX on VisitStatus {
  String get value {
    switch (this) {
      case VisitStatus.pending:
        return 'pending';
      case VisitStatus.inProgress:
        return 'in_progress';
      case VisitStatus.gagalPartial:
        return 'gagal_partial';
      case VisitStatus.completed:
        return 'completed';
      case VisitStatus.gagalTotal:
        return 'gagal_total';
    }
  }

  String get label {
    switch (this) {
      case VisitStatus.pending:
        return 'Pending';
      case VisitStatus.inProgress:
        return 'In Progress';
      case VisitStatus.gagalPartial:
        return 'Gagal Partial';
      case VisitStatus.completed:
        return 'Completed';
      case VisitStatus.gagalTotal:
        return 'Gagal Total';
    }
  }

  static VisitStatus fromString(String s) {
    switch (s) {
      case 'pending':
        return VisitStatus.pending;
      case 'in_progress':
        return VisitStatus.inProgress;
      case 'gagal_partial':
        return VisitStatus.gagalPartial;
      case 'completed':
        return VisitStatus.completed;
      case 'gagal_total':
        return VisitStatus.gagalTotal;
      default:
        return VisitStatus.pending;
    }
  }
}

const scheduleStatuses = VisitStatus.values;

const statusTransitions = <VisitStatus, List<VisitStatus>>{
  VisitStatus.pending: [
    VisitStatus.inProgress,
    VisitStatus.gagalPartial,
    VisitStatus.completed,
    VisitStatus.gagalTotal,
  ],
  VisitStatus.inProgress: [
    VisitStatus.gagalPartial,
    VisitStatus.completed,
    VisitStatus.gagalTotal,
  ],
  VisitStatus.gagalPartial: [VisitStatus.completed, VisitStatus.gagalTotal],
  VisitStatus.completed: [VisitStatus.inProgress],
  VisitStatus.gagalTotal: [],
};

bool canTransition(VisitStatus from, VisitStatus to) {
  return statusTransitions[from]!.contains(to);
}
